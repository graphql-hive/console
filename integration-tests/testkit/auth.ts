import { AuthenticationClient, TokenResponse } from 'auth0';
import { ensureEnv } from './env';

const authenticationApi = new AuthenticationClient({
  domain: ensureEnv('AUTH0_DOMAIN'),
  clientId: ensureEnv('AUTH0_CLIENT_ID'),
  clientSecret: ensureEnv('AUTH0_CLIENT_SECRET'),
});

type UserID = 'main' | 'extra';
const password = ensureEnv('AUTH0_USER_PASSWORD');

const userEmails: Record<UserID, string> = {
  main: ensureEnv('AUTH0_USER_MAIN_EMAIL'),
  extra: ensureEnv('AUTH0_USER_EXTRA_EMAIL'),
};

const tokenResponsePromise: Record<UserID, Promise<TokenResponse> | null> = {
  main: null,
  extra: null,
};

export function authenticate(userId: UserID) {
  if (!tokenResponsePromise[userId]) {
    tokenResponsePromise[userId] = authenticationApi.passwordGrant({
      username: userEmails[userId],
      password,
      audience: `https://${ensureEnv('AUTH0_DOMAIN')}/api/v2/`,
      scope: 'openid profile email offline_access',
      realm: 'Username-Password-Authentication',
    });
  }

  return tokenResponsePromise[userId]!;
}
