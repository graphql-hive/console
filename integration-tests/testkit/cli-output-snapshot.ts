import stripAnsi from 'strip-ansi';
import type { SnapshotSerializer } from 'vitest';
import { ExecaError } from '@esm2cjs/execa';

// ------------------------------
// Value
// ------------------------------

type Value = ValueSansCleanable | Cleanable<ValueSansCleanable>;

type ValueSansCleanable = ValueUnwrapped | PromiseSettledResult<ValueUnwrapped>;

type ValueUnwrapped = string | ExecaError;

// ------------------------------
// Cleanable
// ------------------------------

interface Cleanable<$Value extends ValueSansCleanable> {
  value: $Value;
  clean: Cleaner;
}

type Cleaner = RegExp | ((value: string) => string);

const isCleanable = (value: unknown): value is Cleanable<ValueSansCleanable> => {
  return (
    typeof value === 'object' &&
    value !== null &&
    'value' in value &&
    'clean' in value &&
    (typeof value.clean === 'function' || value.clean instanceof RegExp)
  );
};

const applyValueCleaners = (cleaners: Cleaner[], value: string) => {
  const regexpMatchReplacer = (_: string, ...args: any[]) => {
    const captures = args.slice(0, -2);
    if (captures.length === 0) return '__VAR__';
    if (captures.length === 1) return `${captures[0]}__VAR__`;
    if (captures.length === 2) return `${captures[0]}__VAR__${captures[1]}`;
    console.log(
      'Warning: More than 2 captures in Value Cleaner given to Vitest CLI Output Serializer.',
    );
    return '__VAR__';
  };

  return cleaners.reduce((value, cleaner) => {
    if (cleaner instanceof RegExp) {
      return value.replaceAll(cleaner, regexpMatchReplacer);
    }
    return cleaner(value);
  }, value);
};

// ------------------------------
// Serializer
// ------------------------------

const test = (value: unknown): boolean => {
  if (typeof value === 'string') {
    return variableReplacements.some(replacement => replacement.pattern.test(value));
  }
  if (isCleanable(value)) {
    return test(value.value);
  }
  if (isPromiseSettledFulfilled(value)) {
    return test(value.value);
  }
  if (isPromiseSettledRejected(value)) {
    return test(value.reason);
  }
  return isExecaError(value);
};

/**
 * @remarks The `value` parameter is only type-safe by an implementation of
 * the `test` function which only returns true for matching types of values.
 */
const serialize = (value: Value): string => {
  const text = serialize_({
    value,
    valueCleaners: CliOutputSnapshot.valueCleaners,
  });
  return text + newLine + endDivider;
};

const serialize_ = ({
  value,
  valueCleaners,
}: {
  value: Value;
  valueCleaners: Cleaner[];
}): string => {
  if (isCleanable(value)) {
    const text = serialize_({
      value: value.value,
      valueCleaners: [...valueCleaners, value.clean],
    });
    return text;
  }

  if (isPromiseSettledFulfilled(value)) {
    let text = serialize_({ value: value.value, valueCleaners }) + newLine;
    text += sectionDivider('promise') + newLine;
    text += 'fulfilled';
    return text;
  }

  if (isPromiseSettledRejected(value)) {
    let text = serialize_({ value: value.reason, valueCleaners }) + newLine;
    text += sectionDivider('promise') + newLine;
    text += 'rejected';
    return text;
  }

  //
  // Scalar Cases (non-recursing)
  //
  // When running cleaners, run generic one first so that value cleaners
  // do not have to worry about ANSI codes.
  //

  if (typeof value === 'string') {
    let text = '';
    text += heading('CLI SUCCESS OUTPUT') + newLine;

    text += sectionDivider('stdout') + newLine;
    text += applyValueCleaners(valueCleaners, generalClean(value));

    return text;
  }

  if (isExecaError(value)) {
    let text = '';
    text += heading('CLI FAILURE OUTPUT') + newLine;

    text += sectionDivider('exitCode') + newLine;
    text += value.exitCode + newLine;

    text += sectionDivider('stderr') + newLine;
    text += applyValueCleaners(valueCleaners, generalClean(value.stderr || '__NONE__')) + newLine;

    text += sectionDivider('stdout') + newLine;
    text += applyValueCleaners(valueCleaners, generalClean(value.stdout || '__NONE__'));

    return text;
  }

  return String(applyValueCleaners(valueCleaners, generalClean(value)));
};

/**
 * Strip ANSI codes and mask generally known variables.
 */
const generalClean = (value: string) => {
  // We strip ANSI codes because their output can vary by platform (e.g. between macOS and GH CI linux-based runner)
  // and we don't care enough about CLI output styling to fork our snapshots for it.
  value = stripAnsi(value);
  for (const replacement of variableReplacements) {
    value = value.replaceAll(replacement.pattern, replacement.mask);
  }
  return value;
};

const variableReplacements = [
  {
    pattern: /("reference": "|"requestId": "|"https?:\/\/)[^"]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /"https?:\/\/[^"]+/gi,
    mask: '"__URL__',
  },
  {
    pattern: /(Reference: )[^ ]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /(https?:\/\/)[^\n ]+/gi,
    mask: '$1__URL__',
  },
];

// ------------------------------
// Type Guards
// ------------------------------

/**
 * The esm2cjs execa package we are using is not exporting the error class, so use this.
 */
const isExecaError = (value: unknown): value is ExecaError => {
  // @ts-expect-error
  return typeof value.exitCode === 'number';
};

const isPromiseSettledRejected = (value: unknown): value is PromiseRejectedResult => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'rejected'
  );
};

const isPromiseSettledFulfilled = (value: unknown): value is PromiseFulfilledResult<unknown> => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'fulfilled'
  );
};

// ------------------------------
// Text Helpers
// ------------------------------

const width = 50;
const newLine = '\n';

const heading = (title: string) => {
  const char = ':';
  const gap = ' ';
  const borderSize = Math.max(0, width - title.length - gap.length * 2);
  const borderSizeHalf = borderSize ? Math.round(borderSize / 2) : 0;
  const borderHalf = char.repeat(borderSizeHalf);
  return `${borderHalf}${gap}${title}${gap}${borderHalf}`;
};

const sectionDivider = (title: string) => {
  const char = '-';
  const charEnd = ':';
  const dividerSize = Math.max(0, width - title.length - 1);
  const divider = char.repeat(dividerSize) + charEnd;
  return `${title}${divider}`;
};

const endDivider = ':'.repeat(width);

// ------------------------------
// Generic Helpers
// ------------------------------

const identity = <T>(value: T): T => value;

// ------------------------------
// Interface
// ------------------------------

export namespace CliOutputSnapshot {
  export const serializer: SnapshotSerializer = {
    test,
    serialize,
  };

  /**
   * A list of cleaners that will be applied to all values before they are snapshotted.
   *
   * Any cleaners you add here will apply to all snapshots in your test file.
   *
   * The value will be cleaned with the general cleaner first. So for example ANSI codes will have been stripped already.
   *
   * A cleaner can be a RegExp instance or an arbitrary function that receives and returns a string.
   *
   * #### RegExp Cleaners
   * When using RegExps, note the following.
   *
   * With regards to captured groups:
   *
   * String method `.replaceAll` is used so your RegExp MUST be using `g` flag.
   *
   * With regards to flags:
   * - 0 captures: Replace with `__VAR__`
   * - 1 capture: Replace with `$1__VAR__`
   * - 2 captures: Replace with `$1__VAR__$2`
   * - 3+ captures: Replace with `__VAR__` and log a warning. (Don't do this.)
   */
  export const valueCleaners: Cleaner[] = [];

  /**
   * Wrap a value with a cleaner for just that value.
   *
   * For details about cleaners, see {@link valueCleaners}.
   */
  export const withCleaner = (
    value: ValueSansCleanable,
    valueClean: Cleaner,
  ): Cleanable<ValueSansCleanable> => {
    return {
      value,
      clean: valueClean,
    };
  };
}
