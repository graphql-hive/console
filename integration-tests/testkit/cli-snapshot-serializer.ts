import stripAnsi from 'strip-ansi';
import type { SnapshotSerializer } from 'vitest';
import { ExecaError } from '@esm2cjs/execa';

const test = (value: unknown): boolean => {
  if (typeof value === 'string') {
    return variableReplacements.some(replacement => replacement.pattern.test(value));
  }
  if (isPromiseSettledFulfilled(value)) {
    return test(value.value);
  }
  if (isPromiseSettledRejected(value)) {
    return test(value.reason);
  }
  return isExecaError(value);
};

type Value = string | ExecaError | PromiseSettledResult<string | ExecaError>;

/**
 * @remarks The `value` parameter is only type-safe by an implementation of
 * the `test` function which only returns true for these types of values.
 */
const serialize = (value: Value): string => {
  const text = serialize_(value);
  return text + newLine + endDivider;
};

const serialize_ = (value: Value): string => {
  if (isPromiseSettledFulfilled(value)) {
    return serialize_(value.value) + newLine + sectionDivider('promise') + newLine + 'fulfilled';
  }

  if (isPromiseSettledRejected(value)) {
    return serialize_(value.reason) + newLine + sectionDivider('promise') + newLine + 'rejected';
  }

  if (typeof value === 'string') {
    let text = '';
    text += heading('CLI SUCCESS OUTPUT') + newLine;

    text += sectionDivider('stdout') + newLine;
    text += clean(value);

    return text;
  }

  if (isExecaError(value)) {
    let text = '';
    text += heading('CLI FAILURE OUTPUT') + newLine;

    text += sectionDivider('exitCode') + newLine;
    text += value.exitCode + newLine;

    text += sectionDivider('stderr') + newLine;
    text += clean(value.stderr || '__NONE__') + newLine;

    text += sectionDivider('stdout') + newLine;
    text += clean(value.stdout || '__NONE__');

    return text;
  }

  return String(value);
};

export const cliOutputSnapshotSerializer: SnapshotSerializer = {
  test,
  serialize,
};

const variableReplacements = [
  {
    pattern: /("reference": "|"requestId": "|"https?:\/\/)[^"]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /"https?:\/\/[^"]+/gi,
    mask: '"__URL__',
  },
  {
    pattern: /(Reference: )[^ ]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /(https?:\/\/)[^\n ]+/gi,
    mask: '$1__URL__',
  },
];

/**
 * Strip ANSI codes and mask variables.
 */
const clean = (value: string) => {
  // We strip ANSI codes because their output can vary by platform (e.g. between macOS and GH CI linux-based runner)
  // and we don't care enough about CLI output styling to fork our snapshots for it.
  value = stripAnsi(value);
  for (const replacement of variableReplacements) {
    value = value.replaceAll(replacement.pattern, replacement.mask);
  }
  return value;
};

/**
 * The esm2cjs execa package we are using is not exporting the error class, so use this.
 */
const isExecaError = (value: unknown): value is ExecaError => {
  // @ts-expect-error
  return typeof value.exitCode === 'number';
};

const isPromiseSettledRejected = (value: unknown): value is PromiseRejectedResult => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'rejected'
  );
};

const isPromiseSettledFulfilled = (value: unknown): value is PromiseFulfilledResult<unknown> => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'fulfilled'
  );
};

const width = 50;
const newLine = '\n';

const heading = (title: string) => {
  const char = ':';
  const gap = ' ';
  const borderSize = Math.max(0, width - title.length - gap.length * 2);
  const borderSizeHalf = borderSize ? Math.round(borderSize / 2) : 0;
  const borderHalf = char.repeat(borderSizeHalf);
  return `${borderHalf}${gap}${title}${gap}${borderHalf}`;
};

const sectionDivider = (title: string) => {
  const char = '-';
  const charEnd = ':';
  const dividerSize = Math.max(0, width - title.length - 1);
  const divider = char.repeat(dividerSize) + charEnd;
  return `${title}${divider}`;
};

const endDivider = ':'.repeat(width);
