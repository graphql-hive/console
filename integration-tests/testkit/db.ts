import { sql, DatabasePoolConnectionType } from 'slonik';

export const resetDb = async (conn: DatabasePoolConnectionType) => {
  const migrationTables = ['migrations'];

  const result = await conn.many<{ tablename: string }>(sql`
    SELECT "tablename"
    FROM "pg_tables"
    WHERE "schemaname" = 'public';
  `);

  const tablenames = result.map(({ tablename }) => tablename).filter(tablename => !migrationTables.includes(tablename));

  if (tablenames.length) {
    await conn.query(sql`
      TRUNCATE TABLE 
        ${sql.join(
          tablenames.map(name => sql.identifier([name])),
          sql`,`
        )}
        RESTART IDENTITY
      ;
    `);
  }
};
