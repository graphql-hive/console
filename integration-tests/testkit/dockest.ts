import { DockestService, execa } from 'dockest';
import { containerIsHealthyReadinessCheck, zeroExitCodeReadinessCheck } from 'dockest/dist/readiness-check/index.js';
import { ReadinessCheck } from 'dockest/dist/@types.js';
import { mapTo, take, filter } from 'rxjs/operators/index.js';
import { DepGraph } from 'dependency-graph';
import { readFileSync } from 'fs';
import { join } from 'path';
import { parse } from 'yaml';

export const startReadinessCheck: ReadinessCheck = ({ runner }) => {
  return runner.dockerEventStream$.pipe(
    filter(ev => ev.action === 'start'),
    mapTo(undefined),
    take(1)
  );
};

export function createServices() {
  const dockerComposeFile: {
    services: {
      [key: string]: {
        depends_on?: { [key: string]: unknown };
        healthcheck?: any;
        labels?: {
          'dockest.readiness'?: 'exit_code';
        };
      };
    };
  } = parse(readFileSync(join(process.cwd(), 'docker-compose.yml'), 'utf8'));

  const serviceNameCollection = Object.keys(dockerComposeFile.services);

  const graph = new DepGraph<DockestService>();

  // First, add all services to the graph
  for (const serviceName of serviceNameCollection) {
    const service = dockerComposeFile.services[serviceName];
    graph.addNode(serviceName, {
      serviceName,
      dependsOn: [],
      readinessCheck: service.healthcheck
        ? containerIsHealthyReadinessCheck
        : service?.labels?.['dockest.readiness'] === 'exit_code'
        ? zeroExitCodeReadinessCheck
        : startReadinessCheck,
    });
  }

  // Now, create dependencies between them
  for (const serviceName of serviceNameCollection) {
    const dockerService = dockerComposeFile.services[serviceName];
    if (dockerService.depends_on) {
      const dependsOn = Object.keys(dockerService.depends_on);

      for (const depName of dependsOn) {
        graph.addDependency(serviceName, depName);
      }
    }
  }

  // Next, sort the graph
  const allServices = graph.overallOrder();

  // Finally, create the services
  const registry: {
    [key: string]: DockestService;
  } = {};
  for (const serviceName of allServices) {
    const service = graph.getNodeData(serviceName);

    registry[serviceName] = {
      ...service,
      dependsOn: graph.directDependenciesOf(serviceName).map(dep => graph.getNodeData(dep)),
    };
  }

  // And return a list of services
  return allServices.map(serviceName => graph.getNodeData(serviceName));
}

export function cleanDockerContainers() {
  const output = execa(`docker ps --all --filter "name=integration-tests" --format={{.ID}}:{{.Status}}`);

  if (output.stdout.length) {
    const runningContainers = output.stdout.split('\n');
    for (const line of runningContainers) {
      const [containerId, containerStatus] = line.split(':');
      const containerRunning = containerStatus?.toLowerCase().includes('up');
      if (containerRunning) {
        console.log(`Stopping container ${containerId}`);
        execa(`docker stop ${containerId}`);
      }
      console.log(`Removing container ${containerId} with its volumes`);
      execa(`docker rm -v -f ${containerId}`);
    }

    console.log('Stopped and removed all containers');
  }
}
