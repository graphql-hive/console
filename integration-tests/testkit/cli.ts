import * as utils from 'dockest/test-helper';
import { run } from '../../packages/libraries/cli/src/index';

const registryAddress = utils.getServiceAddress('server', 3001);

export async function schemaPublish(args: string[]) {
  return run(['schema:publish', `--registry`, `http://${registryAddress}/graphql`, ...args]);
}

export async function schemaCheck(args: string[]) {
  return run(['schema:check', `--registry`, `http://${registryAddress}/graphql`, ...args]);
}
