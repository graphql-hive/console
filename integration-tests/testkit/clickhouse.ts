import * as utils from 'dockest/test-helper';
import axios from 'axios';

const clickhouseAddress = utils.getServiceAddress('clickhouse', 8123);
const endpoint = `http://${clickhouseAddress}/?default_format=JSON`;

export async function resetClickHouse() {
  const queries = [
    `operations_registry`,
    `operations_new_hourly_mv`,
    `operations_new`,
    `schema_coordinates_daily`,
    `client_names_daily`,
  ].map(table => `TRUNCATE TABLE default.${table}`);

  for await (const query of queries) {
    await axios.post(endpoint, query, {
      method: 'POST',
      timeout: 10_000,
      headers: {
        'Accept-Encoding': 'gzip',
        Accept: 'application/json',
        Authorization: `Basic ${Buffer.from('test:test').toString('base64')}`,
      },
    });
  }
}

export async function clickHouseQuery<T>(query: string) {
  const res = await axios.post<{
    data: T[];
    rows: number;
  }>(endpoint, query, {
    timeout: 10_000,
    headers: {
      'Accept-Encoding': 'gzip',
      Authorization: `Basic ${Buffer.from('test:test').toString('base64')}`,
    },
    responseType: 'json',
  });

  return res.data;
}
