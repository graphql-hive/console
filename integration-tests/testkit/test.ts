/**
 * This module uses Vitest's fixture system to make common usage patterns
 * of our testkit easily consumable in test cases. @see https://vitest.dev/guide/test-context.html#test-extend
 */

import { test as testBase } from 'vitest';
import { CLI, createCLI } from './cli';
import { ProjectType } from './gql/graphql';
import { initSeed, OrgSeed, OwnerSeed, ProjectSeed, Seed, TargetAccessTokenSeed } from './seed';

interface Context {
  seed: Seed;
  owner: OwnerSeed;
  org: OrgSeed;
  //
  // "single" branch
  //
  projectSingle: ProjectSeed;
  tokenForProjectSingle: TargetAccessTokenSeed;
  cliForProjectSingle: CLI;
  //
  // "federation" branch
  //
  projectFederation: ProjectSeed;
  tokenForProjectFederation: TargetAccessTokenSeed;
  cliForProjectFederation: CLI;
  //
  // "stitching" branch
  //
  projectStitching: ProjectSeed;
  tokenForProjectStitching: TargetAccessTokenSeed;
  cliForProjectStitching: CLI;
}

export const test = testBase.extend<Context>({
  seed: async ({}, use) => {
    const seed = await initSeed();
    await use(seed);
  },
  owner: async ({ seed }, use) => {
    const owner = await seed.createOwner();
    await use(owner);
  },
  org: async ({ owner }, use) => {
    const org = await owner.createOrg();
    await use(org);
  },
  //
  // "single" branch
  //
  projectSingle: async ({ org }, use) => {
    const project = await org.createProject(ProjectType.Single);
    await use(project);
  },
  tokenForProjectSingle: async ({ projectSingle }, use) => {
    const token = await projectSingle.createTargetAccessToken({});
    await use(token);
  },
  cliForProjectSingle: async ({ tokenForProjectSingle }, use) => {
    const cli = createCLI({
      readwrite: tokenForProjectSingle.secret,
      readonly: tokenForProjectSingle.secret,
    });
    await use(cli);
  },
  //
  // "federation" branch
  //
  projectFederation: async ({ org }, use) => {
    const project = await org.createProject(ProjectType.Federation);
    await use(project);
  },
  tokenForProjectFederation: async ({ projectFederation }, use) => {
    const token = await projectFederation.createTargetAccessToken({});
    await use(token);
  },
  cliForProjectFederation: async ({ tokenForProjectFederation }, use) => {
    const cli = createCLI({
      readwrite: tokenForProjectFederation.secret,
      readonly: tokenForProjectFederation.secret,
    });
    await use(cli);
  },
  //
  // "stitching" branch
  //
  projectStitching: async ({ org }, use) => {
    const project = await org.createProject(ProjectType.Stitching);
    await use(project);
  },
  tokenForProjectStitching: async ({ projectStitching }, use) => {
    const token = await projectStitching.createTargetAccessToken({});
    await use(token);
  },
  cliForProjectStitching: async ({ tokenForProjectStitching }, use) => {
    const cli = createCLI({
      readwrite: tokenForProjectStitching.secret,
      readonly: tokenForProjectStitching.secret,
    });
    await use(cli);
  },
});
