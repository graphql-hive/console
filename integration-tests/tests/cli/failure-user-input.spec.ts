import { argsToExecFormat, exec, ExecCommandPath } from '../../testkit/cli';
import { test } from '../../testkit/test';
import { SnapshotSerializers } from './__snapshot_serializers__/__';

expect.addSnapshotSerializer(SnapshotSerializers.cliOutput);

interface TestCase {
  command: ExecCommandPath;
  args?: Record<string, string>;
}

// prettier-ignore
const testCases: TestCase[] = [
  { command: 'whoami' },
  { command: 'schema:publish' },
  { command: 'schema:check' },
  { command: 'schema:delete' },
  { command: 'schema:fetch' },
  { command: 'app:create' },
  { command: 'app:publish' },
];

test.each(testCases)('FailureUserInput - %s', async ({ command, args }) => {
  const preparedCommand = `${command} ${argsToExecFormat(args)}`;
  await expect(exec(preparedCommand)).rejects.toMatchSnapshot('OUTPUT FORMAT: TEXT');
  const preparedCommandJson = `${preparedCommand} --json`;
  await expect(exec(preparedCommandJson)).rejects.toMatchSnapshot('OUTPUT FORMAT: JSON');
});
