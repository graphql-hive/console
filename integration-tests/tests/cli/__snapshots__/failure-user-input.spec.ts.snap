// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FailureUserInput - { command: 'whoami' } > OUTPUT FORMAT: JSON 1`] = `
::::::::::::::: CLI FAILURE OUTPUT :::::::::::::::
exitCode-----------------------------------------:
1
stderr-------------------------------------------:
__NONE__
stdout-------------------------------------------:
{
  "data": {
    "type": "FailureUserInput",
    "parameter": "registry.accessToken"
  },
  "type": "failure",
  "reference": null,
  "suggestions": [],
  "warnings": []
}
promise------------------------------------------:
rejected
::::::::::::::::::::::::::::::::::::::::::::::::::
`;

exports[`FailureUserInput - { command: 'whoami' } > OUTPUT FORMAT: TEXT 1`] = `
::::::::::::::: CLI FAILURE OUTPUT :::::::::::::::
exitCode-----------------------------------------:
2
stderr-------------------------------------------:
 ›   Error: --registry.accessToken is required. For help generating an access
 ›   token, see https://__URL__
 ›   stry-access-tokens
stdout-------------------------------------------:
__NONE__
promise------------------------------------------:
rejected
::::::::::::::::::::::::::::::::::::::::::::::::::
`;
